Add Dockerfile and devcontainer.json

dariusarnold opened the pull request:
This allows both interactive use in IDEs/editors with dev container support as well as building inside the dev container from the command line. The changes are based on https://github.com/kon-foo/InfiniSim/tree/docker-builder-image.

The original PR was extended with the following features:
- ccache and ninja are installed for optional faster builds. Using ninja instead of make speeds up the build from ~ 28 s to 17 s on my machine with 16 threads.
- git and sudo are added for interactive use of the devcontainer
- Store bash history in volume to preserve it across rebuilds
- The non-root user infinitime with sudo password "it" was added and is used to avoid producing build artifacts owned by root
- Split args into CMake generation time and CMake build time args
- Allow passing the build directory in docker run

NeroBurner commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+
+WORKDIR /sources
+# Directory if InfiniTime source code
+ENV INFITIME_DIR="/sources/InfiniTime"
```
Anu reason why using different variable names here?

dariusarnold replied:
Just a typo.

dariusarnold commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+FROM ubuntu:22.04
```
The InfiniTime repo is still on 22.04. I thought since this builds much of the same code, we might want to stay at the same version for compilers etc. to avoid incompatibilities.

I don't mind updating though and the pipeline already uses 24.04 with ubuntu-latest, so there should be no issues. I will update to latest LTS.

dariusarnold commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    python3-venv \
+    && rm -rf /var/cache/apt/* /var/lib/apt/lists/* \
+    && curl -sL https://deb.nodesource.com/setup_${NODE_MAJOR}.x -o nodesource_setup.sh \
+    && bash nodesource_setup.sh \
```
Yes, that would stay in the layer, well spotted. I removed it after the install.

dariusarnold commented:
Thank you for the review @NeroBurner, I think I integrated all your findings.

I also changed the following:
- Change the container user to ubuntu:ubuntu since that one already exists anyway. There were also permisison issues stemming from not using ubuntu since the external user would likely have UID 1000 (at least on single user systems). That would lead to external files mapped into the container being owned by ubuntu, not infinitime, which has UID 1001
- Split installation into three sections to increase layer granularity for caching

NeroBurner reviewed (changes requested):
Thanks for the effort! The dev-container will make it easier for new contributors to quickly jump in, thanks!

Regarding whether or not to use the docker container for the CI action I'm undecided. On the one hand I like to reduce code duplication, which is in favor of using docker container. On the other hand the current workflow is minimal, and I don't know what the runtime impact is of building the container vs doing the installs directly. Additionally I don't know what special runner is needed to create a docker container in CI (as far as I understand you need a docker in docker capable runner to create a docker container in CI).

So I'm generally open for docker in CI, just need a bit of hand holding until I can agree

dariusarnold commented:
I added all review suggestions 
I tested this in CLion CLion 2024.3.4 and VsCode and (had to add a workaround for a bug in CLion), the devcontainer works in both. 
Regarding CI usage, I can experiment in my fork to find out about the runtime impact. It is a separate change anyway.

NeroBurner reviewed (approved):
Thanks for the improvements!

NeroBurner commented on .devcontainer/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+    && chown -R $USERNAME /commandhistory \
+    && echo "$SNIPPET" >> "/home/$USERNAME/.bashrc"
+
+USER ubuntu
```
`podman` runs in userland without the need of a docker-daemon running as root. And as such some permissions are restricted.

When running inside the container (without the `USER ubuntu`) with `podman` the shown user is `root`. But the resulting binaries of the bind mount (the `build` dir) have the host user as owner.

So I'd like to remove the `USER ubuntu` part of the `Dockerfile` and update the `README.md` with `--user $(id -u):$(id -g)` for `docker` users.

could you try that out if that would work (as I don't have `docker` installed)?

dariusarnold replied:
I have tested the following after removing the `USER ubuntu` from the Dockerfile (all without a user `ubuntu` on my local machine):
1. Rebuilding the dev container with VsCode, building InfiniSim inside the container
  -> The binaries are owned by the local user and can be run outside of the container as well.
2. Building InfiniSim from command line with `docker run --rm -it -v ${PWD}:/sources  --user $(id -u):$(id -g) infinisim-build`
  -> Same result as in 1.
3. Build the container in CLion and build InfiniSim in the IDE
  -> Same result as in 1.

Good results and all use cases still work.

NeroBurner replied:
awesome! thanks for checking! 🙇

would you be willing to open a PR and update the readme and docker container? if not it's also OK as you were already very helpful!

dariusarnold replied:
Yes, I can do that.
